src/services/geo-location/geo.location.service.ts

```diff
@@ -0,0 +1,115 @@
+import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
+import { HttpService } from '@nestjs/axios';
+import { AxiosError } from 'axios';
+import { Cache } from 'cache-manager';
+import { GeoLocationCacheMetadata } from '@services/geo-location/geo.location.cache.metadata';
+import {
+  GeoServiceErrorException,
+  GeoServiceLimitReachedException,
+  GeoServiceNotAvailableException,
+} from '@common/exceptions/geo';
+import { LogContext } from '@common/enums';
+import { GeoInformation } from './geo.information';
+import { GeoPluginResponse } from './geo.plugin.response';
+
+const endpoint = 'http://www.geoplugin.net/json.gp?ip=';
+const cacheEntryTtl = 14400; // 4 hours
+const geoServiceCallsKey = 'geo-service-call-limit';
+// https://www.geoplugin.com/faq#i_stopped_getting_responses_from_geoplugin.net
+// the free lookup limit of 120 requests per minute.
+const allowedCallsToService = 120;
+const allowedCallsToServiceWindow = 60;
+
+@Injectable()
+export class GeoLocationService {
```

Comment:
This service is one of the things that would make sense to have unit tests, as there is a flow (algorithm) in it.